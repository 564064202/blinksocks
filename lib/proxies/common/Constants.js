"use strict";Object.defineProperty(exports,"__esModule",{value:true});var NOOP=exports.NOOP=0;var SOCKS_VERSION_V4=exports.SOCKS_VERSION_V4=4;var SOCKS_VERSION_V5=exports.SOCKS_VERSION_V5=5;var METHOD_NO_AUTH=exports.METHOD_NO_AUTH=0;var REQUEST_COMMAND_CONNECT=exports.REQUEST_COMMAND_CONNECT=1;var REQUEST_COMMAND_BIND=exports.REQUEST_COMMAND_BIND=2;var REQUEST_COMMAND_UDP=exports.REQUEST_COMMAND_UDP=3;var ATYP_V4=exports.ATYP_V4=1;var ATYP_DOMAIN=exports.ATYP_DOMAIN=3;var ATYP_V6=exports.ATYP_V6=4;var REPLY_GRANTED=exports.REPLY_GRANTED=90;// 90
var REPLY_SUCCEEDED=exports.REPLY_SUCCEEDED=0;// export const REPLY_FAILURE = 0x01;
// export const REPLY_NOT_ALLOWED = 0x02;
// export const REPLY_NETWORK_UNREACHABLE = 0x03;
// export const REPLY_HOST_UNREACHABLE = 0x04;
// export const REPLY_CONNECION_REFUSED = 0x05;
// export const REPLY_TTL_EXXPIRED = 0x06;
var REPLY_COMMAND_NOT_SUPPORTED=exports.REPLY_COMMAND_NOT_SUPPORTED=7;// export const REPLY_ADDRESS_TYPE_NOT_SUPPORTED = 0x08;
var REPLY_UNASSIGNED=exports.REPLY_UNASSIGNED=255;