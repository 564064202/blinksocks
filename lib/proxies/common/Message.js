"use strict";Object.defineProperty(exports,"__esModule",{value:true});var _createClass=function(){function defineProperties(target,props){for(var i=0;i<props.length;i++){var descriptor=props[i];descriptor.enumerable=descriptor.enumerable||false;descriptor.configurable=true;if("value"in descriptor)descriptor.writable=true;Object.defineProperty(target,descriptor.key,descriptor)}}return function(Constructor,protoProps,staticProps){if(protoProps)defineProperties(Constructor.prototype,protoProps);if(staticProps)defineProperties(Constructor,staticProps);return Constructor}}();function _classCallCheck(instance,Constructor){if(!(instance instanceof Constructor)){throw new TypeError("Cannot call a class as a function")}}var Message=exports.Message=function(){function Message(){_classCallCheck(this,Message)}_createClass(Message,[{key:"toBuffer",value:function toBuffer(){return Buffer.from([])}}],[{key:"parse",value:function parse(){return null}}]);return Message}();