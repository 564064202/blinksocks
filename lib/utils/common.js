'use strict';Object.defineProperty(exports,'__esModule',{value:true});exports.BYTE_ORDER_LE=exports.BYTE_ORDER_BE=exports.ATYP_V6=exports.ATYP_DOMAIN=exports.ATYP_V4=undefined;exports.numberToBuffer=numberToBuffer;exports.parseURI=parseURI;exports.getRandomInt=getRandomInt;exports.getRandomChunks=getRandomChunks;exports.getChunks=getChunks;var _net=require('net');var _net2=_interopRequireDefault(_net);var _crypto=require('crypto');var _crypto2=_interopRequireDefault(_crypto);var _ip=require('ip');var _ip2=_interopRequireDefault(_ip);var _urijs=require('urijs');var _urijs2=_interopRequireDefault(_urijs);function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj}}const ATYP_V4=exports.ATYP_V4=1;const ATYP_DOMAIN=exports.ATYP_DOMAIN=3;const ATYP_V6=exports.ATYP_V6=4;const BYTE_ORDER_BE=exports.BYTE_ORDER_BE=0;const BYTE_ORDER_LE=exports.BYTE_ORDER_LE=1;function numberToBuffer(num,len=2,byteOrder=BYTE_ORDER_BE){if(len<1){throw Error('len must be greater than 0')}const isOutOfRange=num>parseInt(`0x${'ff'.repeat(len)}`);if(isOutOfRange){throw Error(`Number ${num} is too big to put into a ${len} byte(s) size buffer`)}const buf=Buffer.alloc(len);if(byteOrder===BYTE_ORDER_BE){buf.writeUIntBE(num,0,len)}else{buf.writeUIntLE(num,0,len)}return buf}function parseURI(uri){let _uri=uri;let _port=null;if(_uri.startsWith('http://')){_uri=_uri.substr(7);_port=80}if(_uri.startsWith('https://')){_uri=_uri.substr(8);_port=443}const parts={};_urijs2.default.parseHost(_uri,parts);const hostname=parts.hostname,port=parts.port;const addrType=_net2.default.isIP(hostname)?_net2.default.isIPv4(hostname)?ATYP_V4:ATYP_V6:ATYP_DOMAIN;return{type:addrType,host:_net2.default.isIP(hostname)?_ip2.default.toBuffer(hostname):Buffer.from(hostname),port:numberToBuffer(port||_port||80)}}function getRandomInt(min,max){min=Math.ceil(min);max=Math.ceil(max);return Math.floor(_crypto2.default.randomBytes(1)[0]/255*(max-min))+min}function getRandomChunks(buffer,min,max){const totalLen=buffer.length;const bufs=[];let ptr=0;while(ptr<totalLen-1){const offset=getRandomInt(min,max);bufs.push(buffer.slice(ptr,ptr+offset));ptr+=offset}if(ptr<totalLen){bufs.push(buffer.slice(ptr))}return bufs}function getChunks(buffer,maxSize){const totalLen=buffer.length;const bufs=[];let ptr=0;while(ptr<totalLen-1){bufs.push(buffer.slice(ptr,ptr+maxSize));ptr+=maxSize}if(ptr<totalLen){bufs.push(buffer.slice(ptr))}return bufs}