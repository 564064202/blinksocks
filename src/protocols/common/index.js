export * from './Message';
export * from './Constants';
