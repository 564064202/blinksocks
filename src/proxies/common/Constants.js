export const NOOP = 0x00;
export const SOCKS_VERSION_V4 = 0x04;
export const SOCKS_VERSION_V5 = 0x05;
export const METHOD_NO_AUTH = 0x00;

export const REQUEST_COMMAND_CONNECT = 0x01;
export const REQUEST_COMMAND_BIND = 0x02;
export const REQUEST_COMMAND_UDP = 0x03;

export const ATYP_V4 = 0x01;
export const ATYP_DOMAIN = 0x03;
export const ATYP_V6 = 0x04;

export const REPLY_GRANTED = 0x5a; // 90
export const REPLY_SUCCEEDED = 0x00;
// export const REPLY_FAILURE = 0x01;
// export const REPLY_NOT_ALLOWED = 0x02;
// export const REPLY_NETWORK_UNREACHABLE = 0x03;
// export const REPLY_HOST_UNREACHABLE = 0x04;
// export const REPLY_CONNECION_REFUSED = 0x05;
// export const REPLY_TTL_EXXPIRED = 0x06;
export const REPLY_COMMAND_NOT_SUPPORTED = 0x07;
// export const REPLY_ADDRESS_TYPE_NOT_SUPPORTED = 0x08;
export const REPLY_UNASSIGNED = 0xff;
