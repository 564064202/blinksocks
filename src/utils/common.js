import net from 'net';
import crypto from 'crypto';
import ip from 'ip';
import url from 'urijs';

export const ATYP_V4 = 1;
export const ATYP_DOMAIN = 3;
export const ATYP_V6 = 4;

export const BYTE_ORDER_BE = 0;
export const BYTE_ORDER_LE = 1;

/**
 * convert a number to a buffer with specified length in specified byte order
 * @param num
 * @param len
 * @param byteOrder
 * @returns {Buffer}
 */
export function numberToBuffer(num, len = 2, byteOrder = BYTE_ORDER_BE) {
  if (len < 1) {
    throw Error('len must be greater than 0');
  }

  const isOutOfRange = num > parseInt(`0x${'ff'.repeat(len)}`);
  if (isOutOfRange) {
    throw Error(`Number ${num} is too big to put into a ${len} byte(s) size buffer`);
  }

  const buf = Buffer.alloc(len);
  if (byteOrder === BYTE_ORDER_BE) {
    buf.writeUIntBE(num, 0, len);
  } else {
    buf.writeUIntLE(num, 0, len);
  }
  return buf;
}

/**
 * convert a http(s) url to an address with type, host and port
 * @param uri
 * @returns {{type: Number, host: Buffer, port: Buffer}}
 */
export function parseURI(uri) {
  let _uri = uri;
  let _port = null;

  if (_uri.startsWith('http://')) {
    _uri = _uri.substr(7);
    _port = 80;
  }

  if (_uri.startsWith('https://')) {
    _uri = _uri.substr(8);
    _port = 443;
  }

  const parts = {};
  url.parseHost(_uri, parts);

  const {hostname, port} = parts;
  const addrType = net.isIP(hostname) ? (net.isIPv4(hostname) ? ATYP_V4 : ATYP_V6) : ATYP_DOMAIN;

  return {
    type: addrType,
    host: net.isIP(hostname) ? ip.toBuffer(hostname) : Buffer.from(hostname),
    port: numberToBuffer(port || _port || 80)
  };
}

/**
 * returns a random integer in [min, max].
 * @param min
 * @param max
 * @returns {Number}
 */
export function getRandomInt(min, max) {
  min = Math.ceil(min);
  max = Math.ceil(max);
  return Math.floor(crypto.randomBytes(1)[0] / 0xff * (max - min)) + min;
}

/**
 * split buffer into chunks, each chunk size is picked randomly from [min, max]
 * @param buffer
 * @param min
 * @param max
 * @returns {Array<Buffer>}
 */
export function getRandomChunks(buffer, min, max) {
  const totalLen = buffer.length;
  const bufs = [];
  let ptr = 0;
  while (ptr < totalLen - 1) {
    const offset = getRandomInt(min, max);
    bufs.push(buffer.slice(ptr, ptr + offset));
    ptr += offset;
  }
  if (ptr < totalLen) {
    bufs.push(buffer.slice(ptr));
  }
  return bufs;
}

/**
 * split buffer into chunks, the max chunk size is maxSize
 * @param buffer
 * @param maxSize
 * @returns {Array<Buffer>}
 */
export function getChunks(buffer, maxSize) {
  const totalLen = buffer.length;
  const bufs = [];
  let ptr = 0;
  while (ptr < totalLen - 1) {
    bufs.push(buffer.slice(ptr, ptr + maxSize));
    ptr += maxSize;
  }
  if (ptr < totalLen) {
    bufs.push(buffer.slice(ptr));
  }
  return bufs;
}

/**
 * generate random id from [0, max - 1] but exclude ids
 * @param ids
 * @param max
 * @returns {Number}
 */
export function generateMutexId(ids = [], max = 0xff) {
  const seqArr = Array.from(new Array(max).keys());
  for (let i = 0; i < ids.length; ++i) {
    seqArr.splice(ids[i] - i, 1);
  }
  if (seqArr.length > 0) {
    return seqArr[crypto.randomBytes(1)[0] % seqArr.length];
  } else {
    return -1;
  }
}
