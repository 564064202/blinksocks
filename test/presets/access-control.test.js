import path from 'path';
import {
  getPresetClassByName,
  PRESET_FAILED,
  CONNECT_TO_REMOTE,
  CONNECTION_CREATED,
  CONNECTION_CLOSED
} from '../../src/presets';
import {PresetRunner, sleep} from '../common';

test('running on both client and server', async () => {
  const runner = new PresetRunner({
    clazz: getPresetClassByName('access-control'),
    params: {
      acl: path.join(__dirname, 'acl.txt')
    }
  }, {
    __IS_CLIENT__: true,
    __IS_SERVER__: false
  });

  await sleep(20);

  const actionPayload = {
    payload: {
      host: 'example.com',
      port: 443
    }
  };
  runner.notify({type: CONNECT_TO_REMOTE, ...actionPayload});
  runner.notify({type: CONNECTION_CREATED, ...actionPayload});

  expect(await runner.forward('12')).toMatchSnapshot();
  expect(await runner.forward('34')).toMatchSnapshot();

  expect(await runner.backward('56')).toMatchSnapshot();
  expect(await runner.backward('78')).toMatchSnapshot();

  runner.notify({type: PRESET_FAILED});
  runner.notify({type: CONNECTION_CLOSED});
});
