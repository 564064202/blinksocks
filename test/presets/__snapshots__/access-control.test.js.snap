// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`running on both client and server 1`] = `
Object {
  "data": Array [
    49,
    50,
  ],
  "type": "Buffer",
}
`;

exports[`running on both client and server 2`] = `
Object {
  "data": Array [
    51,
    52,
  ],
  "type": "Buffer",
}
`;

exports[`running on both client and server 3`] = `
Object {
  "data": Array [
    53,
    54,
  ],
  "type": "Buffer",
}
`;

exports[`running on both client and server 4`] = `
Object {
  "data": Array [
    55,
    56,
  ],
  "type": "Buffer",
}
`;
